sbi/inference/base.py

```diff
@@ -0,0 +1,128 @@
+            
+            summary_writer: an optional SummaryWriter to control log file 
+                location (default is <current working directory>/logs.)
+             
+            device: torch.device on which to compute (optional).
+            
+            mcmc_method: MCMC method to use for posterior sampling, one of 
+                ['slice', 'hmc', 'nuts'].
+        """
+
+        # check inputs: observation
+        self._true_observation = atleast_2d(true_observation)
+
+        # warn if prior is Uniform and dim_input > 1 (inferred by drawing once)
+        # XXX if prior is a torch Distribution, then it should support
+        # XXX event_shape, no need to sample
+        # XXX also, why use 'input' here? -> parms are input to the simulator
+        scalar_input = prior.sample().numel() == 1
+
+        if isinstance(prior, Uniform) and not scalar_input:
+            warnings.warn(
+                f"""The parameter dimension of the simulator we expect 
+                rom the provided prior's `event_shape` is >1,
+                and the prior is a PyTorch Uniform distribution. 
+                Beware that you are using a `batch_shape` >1 implicitly 
+                and `event_shape` 1, because PyTorch does not support 
+                multivariate Uniform. Please consider using a BoxUniform prior instead."""
+            )
+        # warning if observed data is multidimensional.
+        # XXX we want numel here?
+        if true_observation.squeeze().ndim > 1:
+            warnings.warn(
+                """`true_observation` has several dimensions. 
+                It could be e.g. a matrix (D=2) of observed data or a batch of
+                1D vector observations. SBI currently does not support batches."""
+            )
+
+        # XXX do simulator attributes get used outside of us accessing them from 
+        # XXX inference classes? In that case, we can store them in a dict
+        # XXX here, e.g. self._simspec()...
+        self._simulator = set_simulator_attributes(simulator, prior, true_observation)
+
+        self._prior = prior
+        
+        self._device = get_default_device() if device is None else device
+
+        self._simulation_batch_size = simulation_batch_size
+
+        # initialize roundwise (parameter, observation) storage
+        self._parameter_bank, self._observation_bank = [], []
+
+        # XXX we could instantiate here the Posterior
+        # XXX solve first dispatching to the right PotentialProvider
+        # XXX f"Class name: {self.__class__.__name__}" -> use in dispatching
+        # XXX additional complication now alg_family is not just class name
+        # XXX (distinction introduced between SRE, AALR, not reflected in class)
+        # XXX  could this be done here, instead of in the child class?
+        # XXX  have to pass string or can access subclass name (for alg family)
```

Comment:
1) I think instantiating the posterior here is a good idea.
2) If you want to just merge this PR quickly, you can just use `self.__class__.__name__`: it will not break the code. The only thing that will happen is that there will be a warning for unnormalized densities even for AALR (which is normalized).
3) Then, for the next PR, we could make AALR and SRE separate classes, both in the sre/ directory. This would then fully solve the problem.

Reply:
But I am also happy if we want to leave it like this for now and then make the change once we decide on splitting SRE and AALR into two separate classes.